tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Logging/NLog/DirectSubmission/LogFactoryGetConfigurationForLoggerInstrumentation.cs

```diff
@@ -44,19 +56,105 @@ public class LogFactoryGetConfigurationForLoggerInstrumentation
         /// <param name="name">The name of the logger</param>
         /// <param name="configuration">The logging configuration</param>
         /// <returns>Calltarget state value</returns>
-        public static CallTargetState OnMethodBegin<TTarget, TLoggingConfiguration>(TTarget instance, string name, TLoggingConfiguration configuration)
+        public static CallTargetState OnMethodBegin<TTarget, TLoggingConfiguration>(TTarget instance, string name, ref TLoggingConfiguration configuration)
         {
             var tracerManager = TracerManager.Instance;
+
+            if (!tracerManager.Settings.LogsInjectionEnabledInternal &&
+                !tracerManager.DirectLogSubmission.Settings.IsIntegrationEnabled(IntegrationId.NLog))
+            {
+                return CallTargetState.GetDefault();
+            }
+
+            if (_version is null)
+            {
+                _version = NLogVersionHelper<TTarget>.Version;
+            }
+
+            _ = instance.TryDuckCast<ILogFactoryProxy>(out var logFactoryProxy);
+            _ = instance.TryDuckCast<ILogFactoryPre43Proxy>(out var logFactoryPre43Proxy);
+
+            if (logFactoryPre43Proxy is null)
+            {
+                Log.Warning("Failed to DuckCast the log factory for NLog - Agentless logging and logs injection for NLog won't be functional.");
+                return CallTargetState.GetDefault();
+            }
+
+            // when logging is being stopped/shutdown this instrumentation gets called again
+            // for later versions of NLog, the configuration passed is "null"
+            // but we have an "_isDisposing" field that is set to true in the LogFactory that we can check
+            // for older versions of NLog, the configuration passed is the _previous_ configuration used
+            // and there is no "_isDisposing" field that we can check, so using the ConfigurationTable to keep track
+
+            if (configuration is not null && ConfigurationTable.TryGetValue(configuration, out _))
+            {
+                Log.Debug("We've created a configuration already for NLog - skipping.");
+                return CallTargetState.GetDefault();
+            }
+
+            if (logFactoryProxy is not null && logFactoryProxy.IsDisposing)
+            {
+                // when logging is stopped (e.g., shutdown) the configuration will be set to null
+                // and this instrumentation will get hit again, so to avoid creating a new configuration
+                // and adding the Datadog Direct Submission target we just need to exit here.
+                Log.Debug("NLog LogFactory was being disposed so skipping the logging configuration instrumentations.");
```

Comment:
nit: We should be a bit careful with this number of debug messages - too many customers leave them on all the time, and it slows down startup a lot. I think we can probably remove _most_ of these, as they're mostly useful for us during development 🙂

Reply:
Yeah I had a lot of log messages in here, I think all of them can just be removed

Reply:
Removed